handshake instance and external_instance

Stippe123 opened the pull request:
The lowering from Handshake to Hw is now working for the instructions "handshake.instance" and "handshake.external_instance". In the "handshake.external_instance" the ESIWrap Attribute is still missing

Stippe123 commented:
To clarify: The newly introduces intruction "handshake.external_instance" is very similar to the instruction "handshake.instance". But instead of referencing a handshake function it can reference a hw.module not generated from handshake. Since such a hw.module does not know about the handshaking logic, this needs to be added in a wrapping pass. 
An example lowering looks like this, where the ESIWrap attribute is added by hand. The lowering from Handshake to HW might look like this:
[external_instance.txt](https://github.com/llvm/circt/files/10571830/external_instance.txt)
[lower_external_instance.txt](https://github.com/llvm/circt/files/10571834/lower_external_instance.txt)

Dinistro commented on lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:
```diff
@@ -1059,6 +1063,36 @@ class SelectConversionPattern
+    // check if referenced module already exists
+    hw::HWModuleLike refModule = checkSubModuleOp(ls.parentModule, op);
+    if(!refModule)
+      assert(refModule && "handshake.instance references a module that does not exists");
```
Is this a case that can occur when parsing input IR? If so, you should return failure in that case. An assertion should check preconditions and similar conditions that should hold in *all* cases. If an assertion is triggered, a programmer messed up. If assertions can be reached by providing certain input to an application, something is fishy.

Note that assertions will be eliminated when compiling in release mode, so this could result in undefined behaviour.

If this is indeed something that a verifier should have caught before, rewrite it to 

```
assert(!refModule && "handshake.instance references a module that does not exists");
```
without the if.

Stippe123 replied:
I think this is already checked with the function InstanceOp::verifySymbolUses in the file HandshakeOps.cpp. Therefore I will rewrite it to
`assert(!refModule && "handshake.instance references a module that does not exists");`

Dinistro commented on lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:
```diff
@@ -1782,18 +1816,28 @@ class ExtModuleConversionPattern : public OpConversionPattern<T> {
                              HandshakeLoweringState &ls)
       : OpConversionPattern<T>::OpConversionPattern(typeConverter, context),
         submoduleBuilder(submoduleBuilder), ls(ls) {}
+
```
NIT: drop this whitespace.

Dinistro commented on lib/Dialect/HW/HWAttributes.cpp:
```diff
@@ -332,6 +332,24 @@ void InnerSymAttr::print(AsmPrinter &odsPrinter) const {
+//===----------------------------------------------------------------------===//
+
+
+Attribute ESIWrapAttr::parse(AsmParser &p, Type type) {
```
This can probably be done with an `assemblyFormat` directly in the `.td` file.

Dinistro reviewed (changes requested):
I dropped a some initial comments and let the CI run through.

Make sure to use `clang-format` to ensure that the code you submit is properly formatted. Many editors/IDEs allow you to configure `clang-format` to be executed on save. Alternatively, there are pre-commit hooks that can do the same for you.

Feel free to reach out to me if something is unclear.

Dinistro commented:
Added both Lucas and Morten as reviewers.

mortbopet reviewed (commented):
Thinking about this - do we even _need_ a `handshake.external_instance` operation? (i know this might be a bit confusing since I've advocated for it myself.

@Dinistro added support for empty `handshake.function` declarations, so in theory we should be able to do something like:

```mlir
module {
  // A module implementing @add
  hw.module @add(%a : i32, %b : i32) -> (i32) {
    %0 = arith.addi %a, %b : i32
    handshake.return %0 : i32
  }

  // A predeclaration of the @add_hs function. Some prior pass generated this
  // and attached the attribute to associate with @add
  handshake.func @add_hs(%a : i32, %b : i32) -> (i32) attributes { ESIWrap = {target = @add, type = combinational} }

  handshake.func @main(%a: i32, %b: i32, %c : i32, %d : i32, %ctrl : none) -> (i32, none) {
   // instantiated like any other handshake function
    %0 = handshake.instance @add_hs(%a, %b) : (i32, i32) -> (i32)
    handshake.return %0, %ctrl : i32, i32, none
  }
}
```

this then lowers through HandshakeToHW - importantly, the predeclaration just changed its type to ESI channels


```mlir
  hw.module @add(%a : i32, %b : i32) -> (i32) {
    %0 = arith.addi %a, %b : i32
    handshake.return %0 : i32
  }

  // A predeclaration of the @add_hs function. Some prior pass generated this
  // and attached the attribute to associate with @add
  hw.module @add_hs(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %clock : i1, %reset: i1) -> (out0: !esi.channel<i32>) attributes { ESIWrap = {target = @add, type = combinational} }

  hw.module @main(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %c: !esi.channel<i32>, %d: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i0>) {
    %add_hs.out0 = hw.instance "add_hs" @add_hs(in0: %a: !esi.channel<i32>, in1: %b: !esi.channel<i32>, clock: %clock: i1, reset: %reset: i1) -> (out0: !esi.channel<i32>)
    hw.output %add_hs.out0, %ctrl : !esi.channel<i32>, !esi.channel<i32>, !esi.channel<i0>
  }
```

A follow-up pass will then decide how to elaborate the `hw.module @add_hs`.

mortbopet commented on include/circt/Dialect/HW/HWAttributes.td:
```diff
@@ -299,6 +299,36 @@ def EnumFieldAttr : AttrDef<HWDialect, "EnumField"> {
 }
 
+// An attribute to define the wrapping interface for an external module.
+def ESIWrapAttr : AttrDef<HWDialect, "ESIWrap"> {
```
Do we need a custom attribute for this? Is a dictionary attribute not suitable? I've usually seen custom attributes associated with dialects - however, i suspect that this is going to be a custom attribute/state associated with a set of passes, hence it shouldn't be a dialect attribute.

Stippe123 replied:
I changed it to a DictionaryAttribute with the name esiWrap. It now looks like this:
'''
module {
  hw.module.extern @handshake_external_instance_ext_add(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>) attributes {esiWrap = {target = @ext_add, type = "combinational"}}
  hw.module @ext_add(%a: i32, %b: i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  hw.module @main(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i0>) {
    %handshake_external_instance0.out0 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_add(in0: %a: !esi.channel<i32>, in1: %b: !esi.channel<i32>, clock: %clock: i1, reset: %reset: i1) -> (out0: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %ctrl : !esi.channel<i32>, !esi.channel<i0>
  }
}
'''

Stippe123 replied:
And I completely removed the attribute type hw::ESIWrap since it not used

Stippe123 replied:
```mlir
module {
  hw.module.extern @handshake_external_instance_ext_add(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>) attributes {esiWrap = {target = @ext_add, type = "combinational"}}
  hw.module @ext_add(%a: i32, %b: i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  hw.module @main(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i0>) {
    %handshake_external_instance0.out0 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_add(in0: %a: !esi.channel<i32>, in1: %b: !esi.channel<i32>, clock: %clock: i1, reset: %reset: i1) -> (out0: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %ctrl : !esi.channel<i32>, !esi.channel<i0>
  }
}
```

teqdruid commented on include/circt/Dialect/HW/HWStructure.td:
```diff
@@ -254,7 +254,8 @@ def HWModuleExternOp : HWModuleOpBase<"module.extern"> {
                        ParamDeclArrayAttr:$parameters,
-                       OptionalAttr<StrAttr>:$verilogName);
+                       OptionalAttr<StrAttr>:$verilogName,
+                       OptionalAttr<DictionaryAttr>:$esi_wrap);
```
Drive by comment: why does anything ESI-specific need to be in the `hw` dialect? If you really need to attach something dialect-specific to an hw extern module, please use a dialect attribute.

Stippe123 replied:
So this would mean to add a dialect attribute for example called esiWrap. Should I add this attribute to the dialect ESI or HW?

teqdruid replied:
Sorry, I meant discardable attribute. They're called "dialect attributes" in the API and "discardable attributes" in the documentation. It means they're not listed in the op definition. (And can be dropped by passes.)

RamirezLucas commented on include/circt/Dialect/Handshake/HandshakeOps.td:
```diff
@@ -116,8 +116,9 @@ def FuncOp : Op<Handshake_Dialect, "func", [
-    DeclareOpInterfaceMethods<SymbolUserOpInterface>
+    CallOpInterface, HasClock,
+    DeclareOpInterfaceMethods<SymbolUserOpInterface>,
+
```
nit: drop the empty line

RamirezLucas commented on lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:
```diff
@@ -1059,6 +1064,38 @@ class SelectConversionPattern
+    // Is already checked by the function
+    // handshake::InstanceOp::verifySymbolUses
+    assert(refModule &&
+           "handshake.instance references a module that does not exists");
```
nit: no s to exist

RamirezLucas commented:
> Thinking about this - do we even need a handshake.external_instance operation? (i know this might be a bit confusing since I've advocated for it myself.

I don't have a super strong opinion on this, but I tend to prefer the solution you propose over having an `handshake.external_instance` operation. My reasoning is motivated by the fact that every SSA value has implicit dataflow semantics at the `handshake` level. I think this assumption gets somewhat violated by the `handshake.external_instance` operation because it passes/receives SSA operands to/from a module without such semantics. If dataflow semantics were made explicit in `handshake`, the same program would not typecheck (e.g., we would be passing something like an `!esi.channel<i32>` to a module expecting an `i32` input). While this all remains very trackable and "fixable" by a later pass, I feel we are perhaps unnecessarily introducing an inconsistency in the way we consider SSA values at the `handshake` level. Having a pass running prior to `lower-to-handshake` and creating "wrappers" around `hw.module`s we want to instantiate from `handshake` allows us to achieve the same thing without that inconsistency.

teqdruid commented on include/circt/Dialect/ESI/ESIPasses.td:
```diff
@@ -54,4 +54,11 @@ def LowerESItoHW: Pass<"lower-esi-to-hw", "mlir::ModuleOp"> {
 }
 
+def ESIWrapCombinational: Pass<"esi-wrap-combinational", "mlir::ModuleOp"> {
+  let summary = "Wrap around each hw.external.module with EsiWrap attribute set to combinational";
```
I still don't understand what this pass does. Can you add a test and a description? What do you mean by "set it to combinational"?

It doesn't look like it, but is this just to for ESI not to use SV interfaces on extern modules?

Stippe123 replied:
I will add a better description to the Pass.
This pass enables us to embed an arbitrary purely combinational module (with no notion of valid ready) into a handshaking environment. Here is an example:
We start with a handshake program. As you can see the hw.module @ext_add is purely combinational. We now want to instantiate this module in our handshake program, which we do with the handshake.external_instance instruction. Instead of referencing a handshake.func we now reference a hw.module. (Note: probably this handshake.external instance is not needed since we can do it directly with an empty handshake_function see comment from Morten)
```mlir
module {
  hw.module @ext_add(%a : i32, %b : i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  handshake.func @main(%a: i32, %b: i32, %ctrl : none) -> (i32, none) {
    // Instantiate ext_add
    %0 = handshake.external_instance @ext_add(%a, %b) : (i32, i32) -> (i32)
    handshake.return %0, %ctrl : i32, none
  }
}
```

From here we lower it to HW/ESI with `--lower-handshake-to-hw`. This looks like this. Important to note here is that hw.module has still no notion of handshaking. But the handshake_external_instance_ext_add module will have a handshaking logic, which instantiates the hw.ext_add instance and adds the necessary wrapping logic around it.
```mlir
module {
  hw.module.extern @handshake_external_instance_ext_add(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>) -> (out0: !esi.channel<i32>) attributes {esiWrap = {target = @ext_add, type = "combinational"}}
  hw.module @ext_add(%a: i32, %b: i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  hw.module @main(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i0>) {
    %handshake_external_instance0.out0 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_add(in0: %a: !esi.channel<i32>, in1: %b: !esi.channel<i32>) -> (out0: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %ctrl : !esi.channel<i32>, !esi.channel<i0>
  }
}
```

With the `--esi-wrap-combinational` pass we now materialize this hw.module.external and add all the wrapping and handshaking logic. This then looks like this:
```mlir
module {
  hw.module @handshake_external_instance_ext_add(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>) -> (out0: !esi.channel<i32>) {
    %rawOutput, %valid = esi.unwrap.vr %in0, %0 : i32
    %rawOutput_0, %valid_1 = esi.unwrap.vr %in1, %0 : i32
    %ext_add.out = hw.instance "ext_add" @ext_add(a: %rawOutput: i32, b: %rawOutput_0: i32) -> (out: i32)
    %chanOutput, %ready = esi.wrap.vr %ext_add.out, %0 : i32
    %0 = comb.and %valid, %valid_1, %ready : i1
    hw.output %chanOutput : !esi.channel<i32>
  }
  hw.module @ext_add(%a: i32, %b: i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  hw.module @main(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i0>) {
    %handshake_external_instance0.out0 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_add(in0: %a: !esi.channel<i32>, in1: %b: !esi.channel<i32>) -> (out0: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %ctrl : !esi.channel<i32>, !esi.channel<i0>
  }
}
```
From here CIRCT will take over. 

So why do we want something like this: In the combinational case we could have a latency benefit but with a reduced wall clock frequency due to longer combinational paths. It gets more interesting if we take the same idea and apply it to a statically pipelined module. (I am now working on such a wrapping pass). In this case we could achieve almost the same performance but with potentially area usage savings.

teqdruid replied:
> This pass enables us to embed an arbitrary purely combinational module (with no notion of valid ready) into a handshaking environment.

That makes sense. Since it sounds like a relatively specific use case, I would be more comfortable with this pass being in the handshake dialect then. Also, is using a `hw.module` really appropriate? Since you need to guarantee that the logic is entirely combinational, wouldn't an operation which has that verifier be more appropriate?

> So why do we want something like this: In the combinational case we could have a latency benefit but with a reduced wall clock frequency due to longer combinational paths. It gets more interesting if we take the same idea and apply it to a statically pipelined module. (I am now working on such a wrapping pass). In this case we could achieve almost the same performance but with potentially area usage savings

Take a look at the (relatively new) [`pipeline.pipeline` op](https://circt.llvm.org/docs/Dialects/Pipeline/#pipelinepipeline-circtpipelinepipelineop). It was intended for exactly this. We haven't used it yet for that purpose (or at all) so it may not fit exactly, but I would encourage you to extend/change that instead of re-inventing something.

Coincidentally, the pipeline ops aren't intended to support any state (they start as combinational only) and the state is scheduled. I've always been uncomfortable with the "combinational only" starting condition being implicit. I'd much rather start with a combinational container op being scheduled during a lowering pass into a pipelined operation.

Dinistro commented on lib/Dialect/ESI/ESIPasses.cpp:
```diff
@@ -857,6 +857,250 @@ void ESIPortsPass::updateInstance(HWModuleExternOp mod, InstanceOp inst) {
+  if (esiWrapAttr) {
+    FlatSymbolRefAttr targetAttr =
+        dyn_cast<FlatSymbolRefAttr>(esiWrapAttr.get("target"));
+    if (targetAttr) {
```
NIT: Drop braces.

Dinistro commented on lib/Dialect/ESI/ESIPasses.cpp:
```diff
@@ -857,6 +857,250 @@ void ESIPortsPass::updateInstance(HWModuleExternOp mod, InstanceOp inst) {
+
+  for (auto extMod :
+       llvm::make_early_inc_range(top.getOps<hw::HWModuleExternOp>())) {
+    if (getEsiWrapType(extMod) == "combinational") {
```
NIT: Use a continues instead to reduce the amount of indentations.

Dinistro commented on lib/Dialect/ESI/ESIPasses.cpp:
```diff
@@ -857,6 +857,250 @@ void ESIPortsPass::updateInstance(HWModuleExternOp mod, InstanceOp inst) {
+  // Build Ready valid network
+  auto andValidReady =
+      b.createOrFold<comb::AndOp>(wrappingMod.getLoc(), validReady, false);
+  // // set the backedge
```
NIT: Drop the additional `//`.

Dinistro reviewed (commented):
Dropped a bunch of NIT comments that point out minor coding style issues.

Stippe123 commented:
I added two new wrapping passes. The fist one is called EsiWrapCalyxPipeline and the second one is called EsiWrapPipeline. The EsiWrapCalyxPipeline wrapping pass wraps a generated statically scheduled pipeline from the CIRCT Pipeline dialect (pipeline.pipeline) into a general pipeline interface. The second pass, EsiWrapPipeline, builds a wrapper for this general pipeline interface, such that it can be properly embedded into a dynamic environment. This pass is based on this paper: https://dynamo.ethz.ch/wp-content/uploads/sites/22/2022/06/Cheng_FPGA20_CombiningDynamicAndStaticSchedulingInHigh-levelSynthesis.pdf (Figure 6b). I have run already some software simulations with cocotb for simple programs and everything seems to work.

Here is an example compilation flow with the new wrapping passes and the handshake.external_instance instruction:
```mlir
module {
  hw.module @ext_complex_arith(%arg0: i32, %arg1: i32, %arg2: i32, %go: i1, %clock: i1, %reset: i1) -> (dout0: i32, dout1: i32, ap_done: i1) {
    %dout0, %dout1, %ap_done = pipeline.pipeline(%arg0, %arg1, %arg2, %go) clock %clock reset %reset : (i32, i32, i32, i1) -> (i32, i32, i1) {
    ^bb0(%arg0_0: i32, %arg0_1: i32, %arg0_2: i32,  %arg3: i1):
      %0 = comb.add %arg0_0, %arg0_1 : i32
      %c_10 = hw.constant 10 : i32
      %1 = comb.sub %arg0_2, %c_10 : i32
      %reg_0, %reg_1, %valid = pipeline.stage.register when %arg3 regs %0, %1 : i32, i32
      %2 = comb.add %reg_0, %reg_1 : i32
      %reg_2, %reg_1_, %valid_1 = pipeline.stage.register when %valid regs %2, %reg_1 : i32, i32
      %c_3 = hw.constant 3 : i32
      %c_7 = hw.constant 7 : i32
      %3 = comb.sub %reg_2, %c_3 : i32
      %4 = comb.add %reg_1_, %c_7 : i32
      pipeline.return %3, %4, %valid_1 valid %valid_1 : i32, i32, i1
    }
    hw.output %dout0, %dout1, %ap_done : i32, i32, i1
  }

  handshake.func @main(%arg0: i32, %arg1: i32, %arg2: i32, %ctrl : none) -> (i32, i32, none) {
    %out0, %out1 = handshake.external_instance @ext_complex_arith(%arg0, %arg1, %arg2) {type = "calyx_pipeline"} : (i32, i32, i32) -> (i32, i32)
    handshake.return %out0, %out1, %ctrl : i32, i32, none
  }
}
```
First we run the lowering `--lower-pipeline-to-hw`. Here the latency attribute needs to be added by hand at the moment:
```mlir
module {
  hw.module @ext_complex_arith(%arg0: i32, %arg1: i32, %arg2: i32, %go: i1, %clock: i1, %reset: i1) -> (dout0: i32, dout1: i32, ap_done: i1) attributes {latency = 2} {
    %c-3_i32 = hw.constant -3 : i32
    %c-10_i32 = hw.constant -10 : i32
    %c7_i32 = hw.constant 7 : i32
    %0 = comb.add %arg0, %arg1 : i32
    %1 = comb.add %arg2, %c-10_i32 : i32
    %s0_valid = seq.compreg %go, %clock : i1
    %s0_reg0 = seq.compreg %0, %clock : i32
    %s0_reg1 = seq.compreg %1, %clock : i32
    %2 = comb.add %s0_reg0, %s0_reg1 : i32
    %s1_valid = seq.compreg %s0_valid, %clock : i1
    %s1_reg0 = seq.compreg %2, %clock : i32
    %s1_reg1 = seq.compreg %s0_reg1, %clock : i32
    %3 = comb.add %s1_reg0, %c-3_i32 : i32
    %4 = comb.add %s1_reg1, %c7_i32 : i32
    hw.output %3, %4, %s1_valid : i32, i32, i1
  }
  handshake.func @main(%arg0: i32, %arg1: i32, %arg2: i32, %arg3: none, ...) -> (i32, i32, none) attributes {argNames = ["arg0", "arg1", "arg2", "ctrl"], resNames = ["out0", "out1", "out2"]} {
    %1:2 = external_instance @ext_complex_arith(%arg0, %arg1, %arg2) {type = "calyx_pipeline"} : (i32, i32, i32) -> (i32, i32)
    return %1#0, %1#1, %arg3 : i32, i32, none
  }
}
```
Next we lower everything to HW/ESI with the conversion `--lower-handshake-to-hw`. Here I added the lowering from the handshake.external_instance. Important to note here is, if the externally referenced module is not of type combinational and has more than 1 input, the corresponding input sync module will be generated, which we then later use in the wrapping passes. This sync module is identical to the one that gets generated from the handshake.sync instruction. This synchronization of the inputs is always necessary if we have 2 or more inputs flowing into a static module. I also added the name of the synchronizer into the EsiWrap module for later passes. The code looks like this:
```mlir
module {
  // newly added module.extern for wrapping
  hw.module.extern @handshake_external_instance_ext_complex_arith(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %in2: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>) attributes {esiWrap = {synchronizer = "handshake_sync_in_ui32_ui32_ui32_out_ui32_ui32_ui32", target = @ext_complex_arith, type = "calyx_pipeline"}}
  // corresponding input synchronizer
  hw.module @handshake_sync_in_ui32_ui32_ui32_out_ui32_ui32_ui32(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %in2: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>, out2: !esi.channel<i32>) {
    %rawOutput, %valid = esi.unwrap.vr %in0, %1 : i32
    %rawOutput_0, %valid_1 = esi.unwrap.vr %in1, %1 : i32
    ...
    ...
    %19 = comb.or %18, %emitted_2 {sv.namehint = "done2"} : i1
    %20 = comb.and %7, %13, %19 {sv.namehint = "allDone"} : i1
    hw.output %chanOutput, %chanOutput_4, %chanOutput_6 : !esi.channel<i32>, !esi.channel<i32>, !esi.channel<i32>
  }
  hw.module @ext_complex_arith(%arg0: i32, %arg1: i32, %arg2: i32, %go: i1, %clock: i1, %reset: i1) -> (dout0: i32, dout1: i32, ap_done: i1) attributes {latency = 2 : i64} {
    %c-3_i32 = hw.constant -3 : i32
    %c-10_i32 = hw.constant -10 : i32
    %c7_i32 = hw.constant 7 : i32
    %0 = comb.add %arg0, %arg1 : i32
    %1 = comb.add %arg2, %c-10_i32 : i32
    %s0_valid = seq.compreg %go, %clock : i1
    %s0_reg0 = seq.compreg %0, %clock : i32
    %s0_reg1 = seq.compreg %1, %clock : i32
    %2 = comb.add %s0_reg0, %s0_reg1 : i32
    %s1_valid = seq.compreg %s0_valid, %clock : i1
    %s1_reg0 = seq.compreg %2, %clock : i32
    %s1_reg1 = seq.compreg %s0_reg1, %clock : i32
    %3 = comb.add %s1_reg0, %c-3_i32 : i32
    %4 = comb.add %s1_reg1, %c7_i32 : i32
    hw.output %3, %4, %s1_valid : i32, i32, i1
  }
  hw.module @main(%arg0: !esi.channel<i32>, %arg1: !esi.channel<i32>, %arg2: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>, out2: !esi.channel<i0>) {
    %handshake_external_instance0.out0, %handshake_external_instance0.out1 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_complex_arith(in0: %arg0: !esi.channel<i32>, in1: %arg1: !esi.channel<i32>, in2: %arg2: !esi.channel<i32>, clock: %clock: i1, reset: %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %handshake_external_instance0.out1, %ctrl : !esi.channel<i32>, !esi.channel<i32>, !esi.channel<i0>
  }
}
```
From there we first embed the module ext_complex_arith into a new module ext_complex_arith_wrap_calyx_pipeline, which has a general pipeline interface with the pass `--esi-wrap-calyx-pipeline` This then looks like this:
```mlir
module {
  // newly added module
  hw.module @ext_complex_arith_wrap_calyx_pipeline(%din0: i32, %din1: i32, %din2: i32, %ap_ce: i1, %clock: i1, %reset: i1) -> (dout0: i32, dout1: i32, ap_ready: i1, ap_done: i1) attributes {latency = 2 : i64} {
    %0 = comb.and %clock, %ap_ce : i1
    %ext_complex_arith.dout0, %ext_complex_arith.dout1, %ext_complex_arith.ap_done = hw.instance "ext_complex_arith" @ext_complex_arith(arg0: %din0: i32, arg1: %din1: i32, arg2: %din2: i32, go: %ap_ce: i1, clock: %0: i1, reset: %reset: i1) -> (dout0: i32, dout1: i32, ap_done: i1)
    hw.output %ext_complex_arith.dout0, %ext_complex_arith.dout1, %ap_ce, %ext_complex_arith.ap_done : i32, i32, i1, i1
  }
  // changed target and type information in the EsiWrap attribute
  hw.module.extern @handshake_external_instance_ext_complex_arith_wrap_calyx_pipeline(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %in2: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>) attributes {esiWrap = {synchronizer = "handshake_sync_in_ui32_ui32_ui32_out_ui32_ui32_ui32", target = @ext_complex_arith_wrap_calyx_pipeline, type = "pipeline"}}
  hw.module @handshake_sync_in_ui32_ui32_ui32_out_ui32_ui32_ui32(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %in2: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>, out2: !esi.channel<i32>) {
    %rawOutput, %valid = esi.unwrap.vr %in0, %1 : i32
    ...
    ...
    hw.output %chanOutput, %chanOutput_4, %chanOutput_6 : !esi.channel<i32>, !esi.channel<i32>, !esi.channel<i32>
  }
  hw.module @ext_complex_arith(%arg0: i32, %arg1: i32, %arg2: i32, %go: i1, %clock: i1, %reset: i1) -> (dout0: i32, dout1: i32, ap_done: i1) attributes {latency = 2 : i64} {
    %c-3_i32 = hw.constant -3 : i32
    ...
    ...
    %4 = comb.add %s1_reg1, %c7_i32 : i32
    hw.output %3, %4, %s1_valid : i32, i32, i1
  }
  hw.module @main(%arg0: !esi.channel<i32>, %arg1: !esi.channel<i32>, %arg2: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>, out2: !esi.channel<i0>) {
    %handshake_external_instance0.out0, %handshake_external_instance0.out1 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_complex_arith_wrap_calyx_pipeline(in0: %arg0: !esi.channel<i32>, in1: %arg1: !esi.channel<i32>, in2: %arg2: !esi.channel<i32>, clock: %clock: i1, reset: %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %handshake_external_instance0.out1, %ctrl : !esi.channel<i32>, !esi.channel<i32>, !esi.channel<i0>
  }
}
```
Now we have the general pipeline interface that we know how to wrap (see paper above). We build the wrapping logic with the pass `--esi-wrap-pipeline`. Here a helper shift register gets created and the sync module from the lowering gets instantiated. For the shift register we need the latency information. This then looks like this:
```mlir
module {
  hw.module @ext_complex_arith_wrap_calyx_pipeline(%din0: i32, %din1: i32, %din2: i32, %ap_ce: i1, %clock: i1, %reset: i1) -> (dout0: i32, dout1: i32, ap_ready: i1, ap_done: i1) attributes {latency = 2 : i64} {
    %0 = comb.and %clock, %ap_ce : i1
    %ext_complex_arith.dout0, %ext_complex_arith.dout1, %ext_complex_arith.ap_done = hw.instance "ext_complex_arith" @ext_complex_arith(arg0: %din0: i32, arg1: %din1: i32, arg2: %din2: i32, go: %ap_ce: i1, clock: %0: i1, reset: %reset: i1) -> (dout0: i32, dout1: i32, ap_done: i1)
    hw.output %ext_complex_arith.dout0, %ext_complex_arith.dout1, %ap_ce, %ext_complex_arith.ap_done : i32, i32, i1, i1
  }
  hw.module @valid_tag_shift_register_2stages(%in_valid: i1, %clock: i1, %reset: i1) -> (out_valid: i1) {
    %false = hw.constant false
    %propagate_valid0 = seq.compreg sym @propagate_valid0 %in_valid, %clock, %reset, %false  : i1
    %propagate_valid1 = seq.compreg sym @propagate_valid1 %propagate_valid0, %clock, %reset, %false  : i1
    hw.output %propagate_valid1 : i1
  }
  // newly added module
  hw.module @handshake_external_instance_ext_complex_arith_wrap_calyx_pipeline_wrap_pipeline(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %in2: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>) {
    %handshake_sync.out0, %handshake_sync.out1, %handshake_sync.out2 = hw.instance "handshake_sync" @handshake_sync_in_ui32_ui32_ui32_out_ui32_ui32_ui32(in0: %in0: !esi.channel<i32>, in1: %in1: !esi.channel<i32>, in2: %in2: !esi.channel<i32>, clock: %clock: i1, reset: %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>, out2: !esi.channel<i32>)
    %rawOutput, %valid = esi.unwrap.vr %handshake_sync.out0, %ext_complex_arith_wrap_calyx_pipeline.ap_ready : i32
    %rawOutput_0, %valid_1 = esi.unwrap.vr %handshake_sync.out1, %ext_complex_arith_wrap_calyx_pipeline.ap_ready : i32
    %rawOutput_2, %valid_3 = esi.unwrap.vr %handshake_sync.out2, %ext_complex_arith_wrap_calyx_pipeline.ap_ready : i32
    %ext_complex_arith_wrap_calyx_pipeline.dout0, %ext_complex_arith_wrap_calyx_pipeline.dout1, %ext_complex_arith_wrap_calyx_pipeline.ap_ready, %ext_complex_arith_wrap_calyx_pipeline.ap_done = hw.instance "ext_complex_arith_wrap_calyx_pipeline" @ext_complex_arith_wrap_calyx_pipeline(din0: %rawOutput: i32, din1: %rawOutput_0: i32, din2: %rawOutput_2: i32, ap_ce: %5: i1, clock: %clock: i1, reset: %reset: i1) -> (dout0: i32, dout1: i32, ap_ready: i1, ap_done: i1)
    %0 = comb.and %valid, %valid_1, %valid_3, %ext_complex_arith_wrap_calyx_pipeline.ap_ready : i1
    %1 = comb.and %clock, %5 : i1
    %valid_tag_shift_register_2stages.out_valid = hw.instance "valid_tag_shift_register_2stages" @valid_tag_shift_register_2stages(in_valid: %0: i1, clock: %1: i1, reset: %reset: i1) -> (out_valid: i1)
    %2 = comb.and %ext_complex_arith_wrap_calyx_pipeline.ap_done, %valid_tag_shift_register_2stages.out_valid : i1
    %true = hw.constant true
    %3 = comb.xor %6, %true : i1
    %4 = comb.and %2, %3 : i1
    %5 = comb.xor %4, %true : i1
    %chanOutput, %ready = esi.wrap.vr %ext_complex_arith_wrap_calyx_pipeline.dout0, %2 : i32
    %chanOutput_4, %ready_5 = esi.wrap.vr %ext_complex_arith_wrap_calyx_pipeline.dout1, %2 : i32
    %6 = comb.and %ready, %ready_5 : i1
    hw.output %chanOutput, %chanOutput_4 : !esi.channel<i32>, !esi.channel<i32>
  }
  hw.module @handshake_sync_in_ui32_ui32_ui32_out_ui32_ui32_ui32(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %in2: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>, out2: !esi.channel<i32>) {
    %rawOutput, %valid = esi.unwrap.vr %in0, %1 : i32
    ...
    ...
    %20 = comb.and %7, %13, %19 {sv.namehint = "allDone"} : i1
    hw.output %chanOutput, %chanOutput_4, %chanOutput_6 : !esi.channel<i32>, !esi.channel<i32>, !esi.channel<i32>
  }
  hw.module @ext_complex_arith(%arg0: i32, %arg1: i32, %arg2: i32, %go: i1, %clock: i1, %reset: i1) -> (dout0: i32, dout1: i32, ap_done: i1) attributes {latency = 2 : i64} {
    %c-3_i32 = hw.constant -3 : i32
    ...
    ...
    hw.output %3, %4, %s1_valid : i32, i32, i1
  }
  hw.module @main(%arg0: !esi.channel<i32>, %arg1: !esi.channel<i32>, %arg2: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>, out2: !esi.channel<i0>) {
    %handshake_external_instance0.out0, %handshake_external_instance0.out1 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_complex_arith_wrap_calyx_pipeline_wrap_pipeline(in0: %arg0: !esi.channel<i32>, in1: %arg1: !esi.channel<i32>, in2: %arg2: !esi.channel<i32>, clock: %clock: i1, reset: %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %handshake_external_instance0.out1, %ctrl : !esi.channel<i32>, !esi.channel<i32>, !esi.channel<i0>
  }
}
```
We are done. Now CIRCT takes over!

I don't have any performance (latency, throughput) and area usage information of the generated code yet. But I will work on that. Further I need to finish my bachelor thesis in the next couple of weeks, so my focus will lay there. Afterwards I am looking forward to clean up the code and make it CIRCT ready, that maybe some parts of this work could go upstream in separate PRs to CIRCT :)
